### ISSUE TITLE
Tests fail on python 3.7

### ISSUE THREAD

**Issue by eli-schwartz** [celery/celery#4913]
Arch Linux is currently trying to upgrade to python 3.7, and during the process of rebuilding all packages, celery failed to build with... several types of errors.

After backporting all existing python 3.7 compat commits from master on top of release 4.2.1, no more async errors exist, but the testsuite still fails. Testing with git master confirms that the following errors persist there too:

```
=================================== FAILURES ===================================
_______________ test_asynloop.test_poll_write_generator_stopped ________________

    def Gen():
>       raise StopIteration()
E       StopIteration

t/unit/worker/test_loops.py:375: StopIteration

The above exception was the direct cause of the following exception:

self = <t.unit.worker.test_loops.test_asynloop object at 0x7f64a7528978>

>   ???

t/unit/worker/test_loops.py:383: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
celery/worker/loops.py:91: in asynloop
    next(loop)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Hub@0x7f64a74a9be0: R:0 W:1>, generator = <class 'generator'>
sleep = <built-in function sleep>, min = <built-in function min>
next = <built-in function next>, Empty = <class '_queue.Empty'>
StopIteration = <class 'StopIteration'>, KeyError = <class 'KeyError'>, READ = 1
WRITE = 4, ERR = 24

    def create_loop(self,
                    generator=generator, sleep=sleep, min=min, next=next,
                    Empty=Empty, StopIteration=StopIteration,
                    KeyError=KeyError, READ=READ, WRITE=WRITE, ERR=ERR):
        readers, writers = self.readers, self.writers
        poll = self.poller.poll
        fire_timers = self.fire_timers
        hub_remove = self.remove
        scheduled = self.timer._queue
        consolidate = self.consolidate
        consolidate_callback = self.consolidate_callback
        on_tick = self.on_tick
        propagate = self.propagate_errors
        todo = self._ready
    
        while 1:
            for tick_callback in on_tick:
                tick_callback()
    
            # To avoid infinite loop where one of the callables adds items
            # to self._ready (via call_soon or otherwise), we take pop only
            # N items from the ready set.
            # N represents the current number of items on the set.
            # That way if a todo adds another one to the ready set,
            # we will break early and allow execution of readers and writers.
            current_todos = len(todo)
            for _ in itertools.repeat(None, current_todos):
                if not todo:
                    break
    
                item = todo.pop()
                if item:
                    item()
    
            poll_timeout = fire_timers(propagate=propagate) if scheduled else 1
            #  print('[[[HUB]]]: %s' % (self.repr_active(),))
            if readers or writers:
                to_consolidate = []
                try:
                    events = poll(poll_timeout)
                    #  print('[EVENTS]: %s' % (self.repr_events(events),))
                except ValueError:  # Issue 882
                    raise StopIteration()
    
                for fd, event in events or ():
                    general_error = False
                    if fd in consolidate and \
                            writers.get(fd) is None:
                        to_consolidate.append(fd)
                        continue
                    cb = cbargs = None
    
                    if event & READ:
                        try:
                            cb, cbargs = readers[fd]
                        except KeyError:
                            self.remove_reader(fd)
                            continue
                    elif event & WRITE:
                        try:
                            cb, cbargs = writers[fd]
                        except KeyError:
                            self.remove_writer(fd)
                            continue
                    elif event & ERR:
                        general_error = True
                    else:
                        logger.info(W_UNKNOWN_EVENT, event, fd)
                        general_error = True
    
                    if general_error:
                        try:
                            cb, cbargs = (readers.get(fd) or
                                          writers.get(fd))
                        except TypeError:
                            pass
    
                    if cb is None:
                        self.remove(fd)
                        continue
    
                    if isinstance(cb, generator):
                        try:
>                           next(cb)
E                           RuntimeError: generator raised StopIteration

/usr/lib/python3.7/site-packages/kombu/asynchronous/hub.py:342: RuntimeError
__________________ test_asynloop.test_poll_raises_ValueError ___________________

self = <Hub@0x7f64a73df390: R:1 W:0>, generator = <class 'generator'>
sleep = <built-in function sleep>, min = <built-in function min>
next = <built-in function next>, Empty = <class '_queue.Empty'>
StopIteration = <class 'StopIteration'>, KeyError = <class 'KeyError'>, READ = 1
WRITE = 4, ERR = 24

    def create_loop(self,
                    generator=generator, sleep=sleep, min=min, next=next,
                    Empty=Empty, StopIteration=StopIteration,
                    KeyError=KeyError, READ=READ, WRITE=WRITE, ERR=ERR):
        readers, writers = self.readers, self.writers
        poll = self.poller.poll
        fire_timers = self.fire_timers
        hub_remove = self.remove
        scheduled = self.timer._queue
        consolidate = self.consolidate
        consolidate_callback = self.consolidate_callback
        on_tick = self.on_tick
        propagate = self.propagate_errors
        todo = self._ready
    
        while 1:
            for tick_callback in on_tick:
                tick_callback()
    
            # To avoid infinite loop where one of the callables adds items
            # to self._ready (via call_soon or otherwise), we take pop only
            # N items from the ready set.
            # N represents the current number of items on the set.
            # That way if a todo adds another one to the ready set,
            # we will break early and allow execution of readers and writers.
            current_todos = len(todo)
            for _ in itertools.repeat(None, current_todos):
                if not todo:
                    break
    
                item = todo.pop()
                if item:
                    item()
    
            poll_timeout = fire_timers(propagate=propagate) if scheduled else 1
            #  print('[[[HUB]]]: %s' % (self.repr_active(),))
            if readers or writers:
                to_consolidate = []
                try:
>                   events = poll(poll_timeout)

/usr/lib/python3.7/site-packages/kombu/asynchronous/hub.py:298: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_mock_self = <Mock name='hub.poller.poll' id='140070279374552'>, args = (1.7,)
kwargs = {}

    def __call__(_mock_self, *args, **kwargs):
        # can't use self in-case a function / method we are mocking uses self
        # in the signature
        _mock_self._mock_check_sig(*args, **kwargs)
>       return _mock_self._mock_call(*args, **kwargs)

/usr/lib/python3.7/unittest/mock.py:951: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_mock_self = <Mock name='hub.poller.poll' id='140070279374552'>, args = (1.7,)
kwargs = {}, self = <Mock name='hub.poller.poll' id='140070279374552'>
_new_name = 'poll', _new_parent = None, _call = call(1.7)
seen = {140070562146576}, skip_next_dot = False, do_method_calls = False
name = 'poll', this_mock_call = call.poll(1.7)

    def _mock_call(_mock_self, *args, **kwargs):
        self = _mock_self
        self.called = True
        self.call_count += 1
        _new_name = self._mock_new_name
        _new_parent = self._mock_new_parent
    
        _call = _Call((args, kwargs), two=True)
        self.call_args = _call
        self.call_args_list.append(_call)
        self.mock_calls.append(_Call(('', args, kwargs)))
    
        seen = set()
        skip_next_dot = _new_name == '()'
        do_method_calls = self._mock_parent is not None
        name = self._mock_name
        while _new_parent is not None:
            this_mock_call = _Call((_new_name, args, kwargs))
            if _new_parent._mock_new_name:
                dot = '.'
                if skip_next_dot:
                    dot = ''
    
                skip_next_dot = False
                if _new_parent._mock_new_name == '()':
                    skip_next_dot = True
    
                _new_name = _new_parent._mock_new_name + dot + _new_name
    
            if do_method_calls:
                if _new_name == name:
                    this_method_call = this_mock_call
                else:
                    this_method_call = _Call((name, args, kwargs))
                _new_parent.method_calls.append(this_method_call)
    
                do_method_calls = _new_parent._mock_parent is not None
                if do_method_calls:
                    name = _new_parent._mock_name + '.' + name
    
            _new_parent.mock_calls.append(this_mock_call)
            _new_parent = _new_parent._mock_new_parent
    
            # use ids here so as not to call __hash__ on the mocks
            _new_parent_id = id(_new_parent)
            if _new_parent_id in seen:
                break
            seen.add(_new_parent_id)
    
        ret_val = DEFAULT
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
                raise effect
    
            if not _callable(effect):
                result = next(effect)
                if _is_exception(result):
                    raise result
                if result is DEFAULT:
                    result = self.return_value
                return result
    
>           ret_val = effect(*args, **kwargs)

/usr/lib/python3.7/unittest/mock.py:1017: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (1.7,), kwargs = {}

    def first(*args, **kwargs):
        if not mod or mock.call_count > mod:
            self.close()
>           raise (socket.error() if exc is None else exc)
E           ValueError

t/unit/worker/test_loops.py:109: ValueError

During handling of the above exception, another exception occurred:

self = <Hub@0x7f64a73df390: R:1 W:0>, generator = <class 'generator'>
sleep = <built-in function sleep>, min = <built-in function min>
next = <built-in function next>, Empty = <class '_queue.Empty'>
StopIteration = <class 'StopIteration'>, KeyError = <class 'KeyError'>, READ = 1
WRITE = 4, ERR = 24

    def create_loop(self,
                    generator=generator, sleep=sleep, min=min, next=next,
                    Empty=Empty, StopIteration=StopIteration,
                    KeyError=KeyError, READ=READ, WRITE=WRITE, ERR=ERR):
        readers, writers = self.readers, self.writers
        poll = self.poller.poll
        fire_timers = self.fire_timers
        hub_remove = self.remove
        scheduled = self.timer._queue
        consolidate = self.consolidate
        consolidate_callback = self.consolidate_callback
        on_tick = self.on_tick
        propagate = self.propagate_errors
        todo = self._ready
    
        while 1:
            for tick_callback in on_tick:
                tick_callback()
    
            # To avoid infinite loop where one of the callables adds items
            # to self._ready (via call_soon or otherwise), we take pop only
            # N items from the ready set.
            # N represents the current number of items on the set.
            # That way if a todo adds another one to the ready set,
            # we will break early and allow execution of readers and writers.
            current_todos = len(todo)
            for _ in itertools.repeat(None, current_todos):
                if not todo:
                    break
    
                item = todo.pop()
                if item:
                    item()
    
            poll_timeout = fire_timers(propagate=propagate) if scheduled else 1
            #  print('[[[HUB]]]: %s' % (self.repr_active(),))
            if readers or writers:
                to_consolidate = []
                try:
                    events = poll(poll_timeout)
                    #  print('[EVENTS]: %s' % (self.repr_events(events),))
                except ValueError:  # Issue 882
>                   raise StopIteration()
E                   StopIteration

/usr/lib/python3.7/site-packages/kombu/asynchronous/hub.py:301: StopIteration

The above exception was the direct cause of the following exception:

self = <t.unit.worker.test_loops.test_asynloop object at 0x7f64a74a3da0>

    def test_poll_raises_ValueError(self):
        x = X(self.app)
        x.hub.readers = {6: Mock()}
        poller = x.hub.poller
        x.close_then_error(poller.poll, exc=ValueError)
>       asynloop(*x.args)

t/unit/worker/test_loops.py:419: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <Mock name='obj' id='140070279312440'>
connection = <Mock name='connection' id='140070279312496'>
consumer = <Mock name='consumer' id='140070279312608'>
blueprint = <Mock name='blueprint' id='140070279312720'>
hub = <Hub@0x7f64a73df390: R:1 W:0>
qos = <Mock name='qos' id='140070279312832'>, heartbeat = None
clock = <Mock name='clock' id='140070279312944'>, hbrate = 2.0

    def asynloop(obj, connection, consumer, blueprint, hub, qos,
                 heartbeat, clock, hbrate=2.0):
        """Non-blocking event loop."""
        RUN = bootsteps.RUN
        update_qos = qos.update
        errors = connection.connection_errors
    
        on_task_received = obj.create_task_handler()
    
        _enable_amqheartbeats(hub.timer, connection, rate=hbrate)
    
        consumer.on_message = on_task_received
        obj.controller.register_with_event_loop(hub)
        obj.register_with_event_loop(hub)
        consumer.consume()
        obj.on_ready()
    
        # did_start_ok will verify that pool processes were able to start,
        # but this will only work the first time we start, as
        # maxtasksperchild will mess up metrics.
        if not obj.restart_count and not obj.pool.did_start_ok():
            raise WorkerLostError('Could not start worker processes')
    
        # consumer.consume() may have prefetched up to our
        # limit - drain an event so we're in a clean state
        # prior to starting our event loop.
        if connection.transport.driver_type == 'amqp':
            hub.call_soon(_quick_drain, connection)
    
        # FIXME: Use loop.run_forever
        # Tried and works, but no time to test properly before release.
        hub.propagate_errors = errors
        loop = hub.create_loop()
    
        try:
            while blueprint.state == RUN and obj.connection:
                # shutdown if signal handlers told us to.
                should_stop, should_terminate = (
                    state.should_stop, state.should_terminate,
                )
                # False == EX_OK, so must use is not False
                if should_stop is not None and should_stop is not False:
                    raise WorkerShutdown(should_stop)
                elif should_terminate is not None and should_stop is not False:
                    raise WorkerTerminate(should_terminate)
    
                # We only update QoS when there's no more messages to read.
                # This groups together qos calls, and makes sure that remote
                # control commands will be prioritized over task messages.
                if qos.prev != qos.value:
                    update_qos()
    
                try:
>                   next(loop)
E                   RuntimeError: generator raised StopIteration

celery/worker/loops.py:91: RuntimeError
=============================== warnings summary ===============================
t/unit/app/test_backends.py::test_backends::()::test_get_backend_aliases[amqp://-AMQPBackend]
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_destination_for
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_store_result__no_routing_key
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_mark_as_done
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_pickleable
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_revive
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_is_pickled
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_mark_as_failure
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_expires_is_int
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_expires_is_float
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_expires_is_timedelta
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_store_result_retries
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_poll_no_messages
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_backlog_limit_exceeded
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_poll_result
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_drain_events_decodes_exceptions_in_meta
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_no_expires
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_process_cleanup
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_reload_task_result
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_reload_group_result
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_save_group
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_restore_group
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/backends/test_amqp.py::test_AMQPBackend::()::test_delete_group
  /build/python-celery/src/celery-4.2.1/celery/backends/amqp.py:67: CPendingDeprecationWarning: 
      The AMQP result backend is scheduled for deprecation in     version 4.0 and removal in version v5.0.     Please use RPC backend or a persistent backend.
  
    alternative='Please use RPC backend or a persistent backend.')

t/unit/bin/test_worker.py::test_Worker::()::test_warns_if_running_as_privileged_user
  /build/python-celery/src/celery-4.2.1/celery/platforms.py:796: RuntimeWarning: You're running the worker with superuser privileges: this is
  absolutely not recommended!
  
  Please specify a different user using the --uid option.
  
  User information: uid=0 euid=1000 gid=1000 egid=1000
  
    uid=uid, euid=euid, gid=gid, egid=egid,

t/unit/worker/test_request.py::test_Request::()::test_from_message_invalid_kwargs
  /build/python-celery/src/celery-4.2.1/celery/app/trace.py:561: RuntimeWarning: Exception raised outside body: InvalidTaskError('Task keyword arguments is not a mapping'):
  Traceback (most recent call last):
    File "/build/python-celery/src/celery-4.2.1/celery/app/trace.py", line 359, in trace_task
      kwargs.items
  AttributeError: 'str' object has no attribute 'items'
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/build/python-celery/src/celery-4.2.1/celery/app/trace.py", line 362, in trace_task
      'Task keyword arguments is not a mapping')
  celery.exceptions.InvalidTaskError: Task keyword arguments is not a mapping
  
    exc, exc_info.traceback)))

-- Docs: http://doc.pytest.org/en/latest/warnings.html
======= 2 failed, 2304 passed, 89 skipped, 25 warnings in 50.71 seconds ========
```

**Comment by auvipy** [comment]
python 3.7 support will land in 4.3 release. you can use master to run 3.7. also this seems a pytest 3.4 compat issus which will be fixed soon. use pytest 3.3.x for now.

**Comment by eli-schwartz** [comment]
- I explicitly said I tried this on master and it still failed, hence, I cannot fix anything by using master to run 3.7 as I am already doing so.
- Downgrading from pytest 3.4 to pytest 3.3 will not work, since your testsuite explicitly pins the version of pytest to `pytest>=3.0,<3.3` **and I am already using pytest 3.2 as the testsuite demands**.

I'm entirely bewildered why you closed this issue without any hint that I did anything wrong, one compelling reason to think I'm doing it right (my explicit mention that I did this on master), and some very good reasons in your own code why I would be incapable of doing the wrong thing if I try.

Did you even read what I wrote? Do you even know how your own testsuite operates? I'm honestly not sure what I'm supposed to do here. :disappointed:

**Comment by auvipy** [comment]
in the mean time review the patch https://github.com/celery/celery/pull/4859 and suggest you inputs

**Comment by thedrow** [comment]
@eli-schwartz This is entirely expected.
We only began working on Python 3.7 support and while some fixes are there, others probably aren't.
There's no need for this issue since we'll see those failures ourselves once we introduce Python 3.7 to our build.

**Comment by lelit** [comment]
Another issue is with the presence of the `backends/async.py` module, as `async` is a keyword on Python 3.7+.

**Comment by eli-schwartz** [comment]
> This is entirely expected

I have not said otherwise.

> There's no need for this issue since we'll see those failures ourselves once we introduce Python 3.7 to our build.

That makes no sense and contradicts itself.

I've opened an issue, because there is no other issue, because I thought it would be helpful to let upstream know, and because I want to be notified when it gets fixed.

I will say, that I've opened a bunch of issues for various projects to track python 3.7 support. This (celery) is the first issue where the only response I've gotten is "this is not an issue because we don't officially support 3.7" / "you should not have opened this issue as it is entirely pointless, we'll see the same thing ourselves whenever we get around to introducing python 3.7 to Travis".

Testing it in Travis may let you know of issues, but it definitely doesn't let users and downstream distributors know whether or the issue is know and/or being worked on.

**Comment by eli-schwartz** [comment]
> Another issue is with the presence of the backends/async.py module, as async is a keyword on Python 3.7+.

That is fixed on master... e7002769211f7340f38df80b3112706a8e07cafb

**Comment by auvipy** [comment]
@eli-schwartz I think some misunderstanding is growing. as there is already a PR open supporting 3.7 fully in the matrix this is likely that we will eventually fix the failures, but since we didn't seems to communicate well so we were unable to let everyone clear that we are working towards fixing these 3.7 related issues. So you shared your failures with the hope of notifying and getting them fixed eventually. On my side I admit I initially overlooked full detail of your report so I closed it thinking that they are already fixed. But later whn I cross checked I re-opened it. thanks for reporting anyways

**Comment by eli-schwartz** [comment]
Then why am I being told: "There's no need for this issue since we'll see those failures ourselves"

How exactly is it a **bad thing** to report issues for tracking purposes? I'm basically being told I should not bother to report issues if I think it is likely that the developers will come across the issue on their own sometime in the indeterminate future. :disappointed:

This is really discouraging, and I would be fine with almost any other result whatsoever. I've been looking at plenty of other issues related to our migration to python 3.7, no other project has made me feel snubbed for simply reporting the issue.

**Comment by sudoforge** [comment]
ERROR REPORTING? WE DON'T NEED NO STINKIN' ERRORS!

**Comment by johnarnold** [comment]
@eli-schwartz auvipy's second comment was pointing you at the PR that's already open for fixing the py37 test matrix failures:

https://github.com/celery/celery/pull/4859

Your comments will be most useful there, IMO.

**Comment by eli-schwartz** [comment]
Yes, I'm capable of reading. Also everything I have said thus far here has been regarding the **rest of** the response I have gotten.

> already open for fixing the py37 test matrix failures

Have you even read the PR in question? It's been open for over a month unmerged, and the **only thing it does** is add the matrix in the first place. It fixes nothing, and it's not even a token effort to fix anything. There's quite literally nothing for me to say **about the test matrix itself**.
Test matrixes do not fix things. Test matrixes cause the errors to appear in Travis CI build statuses.

That's very nice but it is also completely unrelated to this very real error that I have already reported by running pytest manually instead of by creating #4859 to run pytest for me.

**Comment by eli-schwartz** [comment]
I would also like to point out that I'm long past caring. I don't use celery, so I'm not invested in the matter. I do use python, and I'm interested in the general python ecosystem as it pertains to my linux distro... and Arch Linux has at this point packaged python 3.7, and our celery package simply ignores all test failures while backporting the two python 3.7 fixes I could find. That's the best I can do, the alternative is simply deleting celery from the repositories, and I'm taking it on faith that it should work -- if any of our users have issues, they're welcome to discuss this more directly themselves.

All that being so, I'm done, I wash my hands of the matter and get back to other projects which interest me on a personal level.

I certainly don't need to spend my time getting reminded a month later (with no upstream activity since then) that I should spend some time navel-gazing about **whether or not to try running unittests in the first place**, for a project that, again, I don't even personally use.

**Comment by carlmjohnson** [comment]
Why was this closed _again_ without a fix? Is there an issue I can follow to know when 3.7 support is released? I can't upgrade my projects to 3.7 until Celery is fixed, so I want to be able to track this fix.
